rust/otap-dataflow/crates/otap/src/compression.rs

```diff
@@ -26,16 +25,73 @@ impl CompressionMethod {
     #[must_use]
     pub const fn map_to_compression_encoding(&self) -> CompressionEncoding {
         match *self {
-            CompressionMethod::Gzip => CompressionEncoding::Gzip,
             CompressionMethod::Zstd => CompressionEncoding::Zstd,
+            CompressionMethod::Gzip => CompressionEncoding::Gzip,
             CompressionMethod::Deflate => CompressionEncoding::Deflate,
         }
     }
 }
 
+/// Default set of compression methods that are accepted when no configuration is provided.
+pub const DEFAULT_COMPRESSION_METHODS: [CompressionMethod; 3] = [
+    CompressionMethod::Zstd,
+    CompressionMethod::Gzip,
+    CompressionMethod::Deflate,
+];
+
+#[derive(Deserialize)]
+#[serde(untagged)]
+enum CompressionConfigValue {
+    Single(CompressionMethod),
+    List(Vec<CompressionMethod>),
+    NoneKeyword(CompressionNone),
+}
+
+#[derive(Deserialize)]
+#[serde(rename_all = "snake_case")]
+enum CompressionNone {
+    None,
+}
+
+/// Deserializer that accepts either a single compression method, a list, or the string `"none"`.
+/// Absence of the field keeps the default behaviour (all methods).
+pub fn deserialize_compression_methods<'de, D>(
+    deserializer: D,
+) -> Result<Option<Vec<CompressionMethod>>, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let value = Option::<CompressionConfigValue>::deserialize(deserializer)?;
+    let Some(value) = value else {
+        return Ok(None);
+    };
+
+    let methods = match value {
+        CompressionConfigValue::Single(method) => vec![method],
+        CompressionConfigValue::List(methods) => methods,
+        CompressionConfigValue::NoneKeyword(CompressionNone::None) => Vec::new(),
+    };
+
+    let mut deduped = Vec::with_capacity(methods.len());
+    for method in methods {
+        if !deduped.contains(&method) {
+            deduped.push(method);
+        }
+    }
```

Comment:
This logic is only needed for `CompressionConfigValue::List`. We could move it inside the match statement to make the code simpler.